Add ability to copy cases between mobile workers for load testing (FF)

Charl1996 opened the pull request:
## Product Description
This PR adds a new page to the UI (Data -> Edit Data) called `Copy Cases`, which aims to simplify the process of accurately assigning cases to load-test users. This feature sits behind a feature flag. 

Below is a screenshot of what this page looks like.

![image](https://github.com/dimagi/commcare-hq/assets/44245062/9434e2ce-000e-48f9-b07e-781958f53d90)

This page will allow the user to copy cases to any other mobile worker whilst de-identifying the user-selected properties. This page is closely modelled after the `CaseReassignmentInterface` page.

Upon successful copy a banner appears at the top of the page indicating the amount of cases copied (this will be equal to the amount of selected case from the case list). On each selected case a notification will also appear indicating that that particular case has been copied. See screenshot below.  

![image](https://github.com/dimagi/commcare-hq/assets/44245062/7b9871ef-de9a-4c7f-8e1f-eae36092d04b)

Copied cases are deleted after the respective owners are deactivated (see [this PR](https://github.com/dimagi/commcare-hq/pull/33353) for more details).

## Technical Summary
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-2636)

The page is modelled after the Reassign Cases page as per the spec. As such, the new interface class, `CaseCopyInterface`, is simply a subclass of `CaseReassignmentInterface` and uses the same base UI page (case_management.html), but with some small tweaks to distinguish the two.

A new "filter" is added, `SensitiveCaseProperties`, which allows the user to select which case properties they want to de-identify. The UI was modelled after the case export UI where a user selects case properties for de-identification.

Note: The `SensitiveCaseProperties` "filter" is not really a filter that will influence the cases appearing in the case list on the page, but rather the selected "filter" values only comes into play in the back-end when the cases are being processed. This is made clear to the user by means of help text below the filter item.

Furthermore, the `copy_cases` method was chosen to reside on the `CaseHelper` class, since this class _"aims to offer a simple interface for simple operations on cases"_. If folks think this should change I'm open to suggestions.

#### What about case relationships?
When the set of cases is copied the copy process aims to keep the integrity of the original case structure, meaning that if a parent-child case is copied, their copied counterparts will also have this relationship. There is one exception, though: when only one of the two cases in this relationship is selected for copying, only that case will be copied. The copied case will thus not have this relationship defined.

## Feature Flag
Copy Cases

## Safety Assurance

### Safety story
- Tested locally
- Included unit tests
- Planned QA on staging

### Automated test coverage
The automated tests cover the following:
- The view:
   -> Access
   -> Parameters
- The copy_cases method on the CaseHelper class
- The 2 new util functions

### QA Plan
[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-5187)

Migrations
 - [ ] The migrations in this code can be safely applied first independently of the code
The migration in this PR is dependant on the code in this PR (adding a new privilege to Pro plan +)

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Charl1996 committed:
Add Copy Cases page

Charl1996 committed:
Link up UI with BE

Charl1996 committed:
Copy cases in backend

Charl1996 committed:
Update UI on case copy success

Charl1996 committed:
Add some tests

Charl1996 committed:
Handle error on UI

Charl1996 committed:
Allow use to specify sensitivity of properties and deid with existing methods

Charl1996 committed:
Add some more tests

Charl1996 committed:
Add help text UI

Charl1996 committed:
Fix stickler issues

Charl1996 committed:
Fix failing test

Charl1996 committed:
Case copy can only be made to another user

kaapstorm commented on corehq/apps/hqcase/case_helper.py:
```diff
@@ -163,6 +167,112 @@ def close(
+            raise Exception('Must copy cases to valid new owner')
+
+        original_cases = CommCareCase.objects.get_cases(case_ids=case_ids, domain=domain)
+        if not any(original_cases):
```
```suggestion
        if not original_cases:
```
... `any()` will check the truthiness of the first case, and I don't think you care about that.

Charl1996 replied:
@kaapstorm What do you mean it "will check the truthiness of the first case"?

As far as I know it checks the truthiness of any item in the iterable, which would in this case either all be truthy or the list would be empty.

(I'm happy to change it, but I can't say I know why)

kaapstorm replied:
Maybe an example:
```
>>> bool([])
False
>>> bool([0])
True
```
As you know, a list is truthy if it has items in it, regardless of the truthiness of the items.
```
>>> any([0])
False
>>> any([0, 1])
True
>>> all([0, 1])
False
>>> all([1])
True
```
`any()` and `all()` evaluate the truthiness of the items inside the list.

A case, as an object or a non-empty dictionary, will always be truthy, so `any()` will return `True` as soon as it encounters the first case. But we don't really need to check whether the case is truthy---we know it is---we just need to check that the list is not empty.
```
>>> case = object()
>>> any([case])
True
>>> bool([case])
True
```

kaapstorm reviewed (commented):
This is great.

A few comments, nothing major.

kaapstorm commented:
Nice tests! :chess_pawn: :pirate_flag:

Charl1996 committed:
Parse docstrings

Charl1996 committed:
Translate text

Charl1996 committed:
Merge branch 'master' into cs/SC-2636-copy_cases_to_test_user

Charl1996 commented:
Update: QA will only be able to test this work next week.

Charl1996 committed:
Display cases based on owner filter

Charl1996 committed:
Merge branch 'master' into cs/SC-2636-copy_cases_to_test_user

Charl1996 committed:
Only have mobile workers as dropdown options

sravfeyn commented on corehq/apps/data_interfaces/interfaces.py:
```diff
@@ -141,6 +145,43 @@ def bulk_response(self):
+    @memoized
+    def es_results(self):
+        query = self._build_query()
+        owner_id = self.request.GET.get('individual')
```
Where is this GET param being used? 

nit: any reason to not call the GET param `owner_id`?

Charl1996 replied:
@sravfeyn 
> Where is this GET param being used?

I'm not sure I know exactly what you're asking?

sravfeyn commented on corehq/apps/data_interfaces/interfaces.py:
```diff
@@ -141,6 +145,43 @@ def bulk_response(self):
+            'corehq.apps.reports.filters.users.SelectMobileWorkerFilter',
+            'corehq.apps.reports.filters.select.MultiCaseTypeFilter',
+            'corehq.apps.reports.standard.cases.filters.CaseSearchFilter',
+            'corehq.apps.reports.standard.cases.filters.SensitiveCaseProperties',
```
nit: You could use `super.fields + ['corehq.apps.reports.standard.cases.filters.SensitiveCaseProperties']` to be DRY

Charl1996 replied:
Ah, right. This is nice

sravfeyn commented on corehq/apps/hqcase/case_helper.py:
```diff
@@ -163,6 +167,110 @@ def close(
+                    new_case_index[identifier] = identifier_index
+            return new_case_index
+
+        def _process_case(_case):
```
This feature can be modelled as a Case Import?

Have you considered using the case-importer related code here directly instead of writing this afresh? This would help keep the case-modification logic in a single place making this, case-import and case-reassignment consistent

Charl1996 replied:
Update: I will start looking into modelling this as a special purpose case import.

Charl1996 replied:
@sravfeyn 
> Have you considered using the case-importer related code here directly instead of writing this afresh? This would help keep the case-modification

It looks like there's a lot that will need changing to make this compatible with the case importer code and/or vice versa. 
That said, I do think it might be helpful to have an abstract class somewhere that handles this type of operation which both this code and the importer code can simply use or inherit from, so I'll create a tech debt ticket to investigate that.

sravfeyn commented on corehq/apps/hqcase/utils.py:
```diff
@@ -238,3 +240,61 @@ def get_last_non_blank_value(case, case_property):
+def get_censored_case_data(case: CommCareCase, censor_data: dict):
+    """
+    This function is used to get the data on the specified case, but with the `censor_data `
+    properties censored.
```
Ah, probably not!

Charl1996 replied:
Mmm...I see what you mean. I'll maybe think of another name.

sravfeyn commented on corehq/apps/hqcase/case_helper.py:
```diff
@@ -1,6 +1,10 @@
 
 from corehq.apps.users.models import CouchUser
+from corehq.form_processor.models.cases import CommCareCase
+from corehq.apps.hqcase.utils import submit_case_blocks, get_censored_case_data
```
nit: `deidentified_data` seems to be more appropriate for the given context over `censored_data`, also consistent with the UI language.

Charl1996 committed:
Merge branch 'master' into cs/SC-2636-copy_cases_to_test_user

Charl1996 committed:
Rename function

Charl1996 commented on corehq/apps/hqcase/case_helper.py:
```diff
@@ -163,6 +167,110 @@ def close(
+                raise Exception("Cannot copy case to self")
+            _process_case(c)
+
+        _, cases = submit_case_blocks(
```
Good point. I'll check it out

Charl1996 committed:
Chunk case submission

Charl1996 committed:
Put case copy feature on Pro plan +

Charl1996 committed:
Merge branch 'master' into cs/SC-2636-copy_cases_to_test_user

Charl1996 committed:
Fix linting issues

Charl1996 committed:
Fix another linting bug

Charl1996 committed:
Add migration lock

Charl1996 committed:
Copy case done by system user

Charl1996 committed:
Fix linting issue

Charl1996 committed:
Fix bug: check name and case_name for name deid

Charl1996 committed:
Fix migration

Charl1996 committed:
Exclude date_opened from sensitive props

SmittieC commented on corehq/apps/reports/filters/api.py:
```diff
@@ -119,10 +128,13 @@ def options_controller(self):
     @property
     @memoized
     def options_controller(self):
+        from corehq.apps.data_interfaces.interfaces import CaseCopyInterface
```
I wonder if we shouldn't do a copy cases privilege check here as well?

Charl1996 replied:
Yeah, good idea.

SmittieC commented on corehq/apps/hqcase/case_helper.py:
```diff
@@ -163,6 +168,128 @@ def close(
+        processed_cases = {}
+        copied_cases_case_blocks = []
+
+        def _get_new_identifier_index(case_identifier_index):
```
Toughts on moving these methods outside the function?

Charl1996 replied:
@SmittieC I created it as nested functions because the functions are by nature very coupled to the copy method and is not useful outside of the copy method. Other than that I don't have strong sentiments for keeping it nested. I wonder what others think?

Charl1996 committed:
Check domain privilege

Charl1996 committed:
Fix merge conflicts with master

Charl1996 committed:
Merge branch 'master' into cs/SC-2636-copy_cases_to_test_user

dannyroberts commented:
Converting to draft based on the sense that this is still being iterated on internally and the informal PR title. Apologies if I'm misreading.

One piece of feedback I have that at first blush this feature feels a bit niche to be introducing so visibly right away to all users on a given plan. Would you be willing to begin by releasing it behind a flag? While its functionality that could be useful to a broader audience eventually, I think we would want to revisit some of its placement, visibility, text, etc to make it really clear that it's a testing utility, not a data management utility like the other things it's currently next to. Without that being really clear I could see this having potential for creating confusion. I'm not super deep in this like you are, so please do let me know if this is something that's already been thought through and include any important context around that decision in the description.

Charl1996 committed:
Add copy cases feature behind feature flag

Charl1996 committed:
Merge branch 'master' into cs/SC-2636-copy_cases_to_test_user

zandre-eng commented on corehq/reports.py:
```diff
@@ -272,7 +272,10 @@ def EDIT_DATA_INTERFACES(domain_obj):
-    if domain_has_privilege(domain_obj.name, privileges.CASE_COPY):
+    if (
+        toggles.COPY_CASES.enabled(domain_obj.name) and
+        domain_has_privilege(domain_obj.name, privileges.CASE_COPY)
```
If we are putting this behind a feature flag, should it be a privilege as well?

Charl1996 replied:
Yeah, it's one of the [requirements](https://dimagi-dev.atlassian.net/browse/SC-2636?focusedCommentId=269252). 

Although, that was before we decided to put it behind a FF. I'll just follow up with the necessary parties about this point.

SmittieC reviewed (approved):
Looks fine to me. Might be good to follow up on [this comment](https://github.com/dimagi/commcare-hq/pull/33054/files#r1230724588) though.

kaapstorm committed:
♻ `CaseHelper.copy_cases()` to `api._copy_cases()`

kaapstorm committed:
♻ `CaseHelper.copy_cases()` to `api._copy_cases()`

Charl1996 committed:
Add to advanced plan

Charl1996 commented:
Last minute change: as per [this comment](https://dimagi-dev.atlassian.net/browse/SC-2636?focusedCommentId=279364) the feature should sit on advanced plan +.

SmittieC reviewed (approved):
Just make sure it's updated on Staging as well

kaapstorm committed:
Drop `count_only` param

kaapstorm committed:
Drop `count_only` param

kaapstorm committed:
Add tests

kaapstorm committed:
Add tests

kaapstorm committed:
Drop some parentheses

kaapstorm committed:
Drop some parentheses

kaapstorm committed:
Use const instead of `function.__name__`

Rather use the const that already exists, than use `function.__name__`
as a const. `DEID_TRANSFORM_FUNCTIONS` shows that these consts are
already being used to look up the transform functions. We should either
stick with that, or we should change the current usage to use
`function.__name__` instead, and the second option looks impractical.

kaapstorm committed:
Use const instead of `function.__name__`

Rather use the const that already exists, than use `function.__name__`
as a const. `DEID_TRANSFORM_FUNCTIONS` shows that these consts are
already being used to look up the transform functions. We should either
stick with that, or we should change the current usage to use
`function.__name__` instead, and the second option looks impractical.

kaapstorm committed:
Drop properties that return consts

Now that we are just using consts, these properties aren't adding much
value.

kaapstorm committed:
Drop properties that return consts

Now that we are just using consts, these properties aren't adding much
value.

kaapstorm committed:
Give `k`, `v` more explanatory names

kaapstorm committed:
Give `k`, `v` more explanatory names

kaapstorm committed:
Re-attach comment to its code

kaapstorm committed:
Re-attach comment to its code

kaapstorm committed:
Apply style guide to type hints, rst to docstrings

* [Style guide](https://docs.google.com/document/d/1q90SaGuxV9dKnjnEIDdu8ma7xrZJxJq_HsfkCUhQqqk/edit)
* [reStructuredText](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html)

kaapstorm committed:
Apply style guide to type hints, rst to docstrings

* [Style guide](https://docs.google.com/document/d/1q90SaGuxV9dKnjnEIDdu8ma7xrZJxJq_HsfkCUhQqqk/edit)
* [reStructuredText](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html)

kaapstorm committed:
Drop type hints

kaapstorm committed:
Drop type hints

kaapstorm committed:
Squash class hierarchy

`_Importer` is never used. YAGNI

kaapstorm committed:
Squash class hierarchy

`_Importer` is never used. YAGNI

kaapstorm committed:
Two ways to import exceptions --> one way

kaapstorm committed:
Two ways to import exceptions --> one way

kaapstorm committed:
`do_import.py`: Add linebreaks and docstrings

kaapstorm committed:
`do_import.py`: Add linebreaks and docstrings

kaapstorm committed:
`do_import.py`: Add type hints

kaapstorm committed:
`do_import.py`: Add type hints

kaapstorm committed:
Add `add_inferred_props_to_schema` param

I chose this approach to perserve current behavior, but I don't think
it's the right one. Instead of using a `soft_assert` in
`submit_and_process_caseblocks()` shouldn't we just check whether
`case_type` is set, and if not, don't try to add properties to its
schema?

kaapstorm committed:
Add `add_inferred_props_to_schema` param

I chose this approach to perserve current behavior, but I don't think
it's the right one. Instead of using a `soft_assert` in
`submit_and_process_caseblocks()` shouldn't we just check whether
`case_type` is set, and if not, don't try to add properties to its
schema?

kaapstorm committed:
`copy_cases()` uses `SubmitCaseBlockHandler`

kaapstorm committed:
`copy_cases()` uses `SubmitCaseBlockHandler`

kaapstorm committed:
Refactor `_copy_cases()` to `CaseCopier` class

* Changes `original_cases` to a dict to make retrieving cases faster.
* Uses IndexAttrs to make it clearer when an index is a dict and when
  it is a tuple.
* Adds some tests for `CaseCopier` methods

kaapstorm committed:
Refactor `_copy_cases()` to `CaseCopier` class

* Changes `original_cases` to a dict to make retrieving cases faster.
* Uses IndexAttrs to make it clearer when an index is a dict and when
  it is a tuple.
* Adds some tests for `CaseCopier` methods

kaapstorm committed:
Drop type hints

kaapstorm committed:
Drop type hints

Charl1996 committed:
Add property to indicate case is a copy

Charl1996 committed:
Restrict custom prop name to be same as copy case prop

Charl1996 committed:
Check for reserved case prop on data dictionary property save

Charl1996 committed:
Label copied cases as such on UI

Charl1996 committed:
Remove copied cases on couch user deactivation

Charl1996 committed:
Use CaseHelper to close cases

Charl1996 committed:
Amend tests

Charl1996 committed:
Add more tests

kaapstorm committed:
Merge pull request #33326 from dimagi/nh/copy_import

Refactor `copy_cases` to use `SubmitCaseBlockHandler`

Charl1996 committed:
Merge branch 'cs/SC-2636-copy_cases_to_test_user' into cs/SC-2904-mark-test-cases

Charl1996 committed:
Check for mobile worker before invoking celery task

Charl1996 committed:
Merge branch 'master' into cs/SC-2636-copy_cases_to_test_user

Charl1996 committed:
Remove test cases on deactivate

Charl1996 committed:
Store original case id in copy prop and move const to class

Charl1996 committed:
Merge branch 'cs/SC-2904-mark-test-cases' into cs/SC-2636-copy_cases_to_test_user

Charl1996 committed:
Update test for clarification

Charl1996 committed:
Merge pull request #33353 from dimagi/cs/SC-2904-mark-test-cases-as-copy

Mark copied cases as "copy" and remove on user deactivation

Charl1996 committed:
Fix failing test - add parent case to submission handler

Charl1996 commented:
Will fix the linting issues in different PR, so consider that non-blocking.

Charl1996 committed:
Fix linting issues

Charl1996 committed:
Rename alert_user to alertUser

Charl1996 committed:
Merge pull request #33362 from dimagi/cs/SC-2636-fix-linting

Fix linting issues

Charl1996 committed:
Ignore F401

Charl1996 committed:
Merge branch 'cs/SC-2636-copy_cases_to_test_user' of github.com:dimagi/commcare-hq into cs/SC-2636-copy_cases_to_test_user

SmittieC commented:
@Charl1996 Just to confirm, this PR will go through a final QA pass?

Charl1996 committed:
Add Copied Case label to CLE

Charl1996 committed:
Merge branch 'master' into cs/SC-2636-copy_cases_to_test_user

Charl1996 commented on corehq/apps/es/case_search.py:
```diff
@@ -420,9 +420,10 @@ def wrap_case_search_hit(hit, include_score=False):
     from `case_json`, even if they were present in the original case's
-    dynamic properties.
+    dynamic properties, except of the COMMCARE_CASE_COPY_PROPERTY_NAME
+    property.
```
`COMMCARE_CASE_COPY_PROPERTY_NAME` should be a reserved property on a case, but it should still show up when wrapping the case result from CaseSearchES since it's needed to determine whether the particular case is a copy.

esoergel replied:
Does it belong in the `SPECIAL_CASE_PROPERTIES` list, then?  Looking at that definition I see
https://github.com/dimagi/commcare-hq/blob/77c2ca5dcf6f5aa02e51da2523b44f5bec59aa31/corehq/apps/case_search/const.py#L38-L41
So it looks like it's for things to be copied from the case object into the set of case properties when sending cases to the case search index.  Since this property is a normal case property that's already there, I'm not sure it's a good fit.

What do you mean by a reserved property?  One that can't be set by users (we can't control what they do in forms, unfortunately)?  One that they can't search for in case search and the CLE?  One that's hidden in reports?

Charl1996 committed:
Merge branch 'master' into cs/SC-2636-copy_cases_to_test_user

Charl1996 committed:
Merge branch 'master' into cs/SC-2636-copy_cases_to_test_user

Charl1996 committed:
Remove commcare_case_copy from SPECIAL_CASE_PROPERTIES_MAP and only onto SPECIAL_CASE_PROPERTIES

zandre-eng reviewed (approved):
Looks good from my side, given that this has passed QA and the PRs that got merged into here have all been approved individually. Only a few minor comments, though none blocking.

zandre-eng commented on corehq/apps/data_interfaces/static/data_interfaces/js/case_management.js:
```diff
@@ -117,6 +131,57 @@ hqDefine("data_interfaces/js/case_management",[
+                    contentType: "application/json",
+                    success: function (response) {
+                        updateCaseRowCopy(self.selectedCases())();
+                        var message = gettext("Successfully copied " + response.copied_cases + " case(s).");
```
@Charl1996 Just flagging that this still needs to be updated.

Charl1996 replied:
I haven't managed to get it to work properly, so I just changed the wording altogether.

zandre-eng commented on corehq/apps/reports/filters/api.py:
```diff
@@ -197,3 +223,194 @@ def get(self, request, domain):
+
+
+@dataclass
+class UserDuck:
```
This class is also defined [here](https://github.com/dimagi/commcare-hq/blob/96044ac2d2d2beddf67a097c7a0cbc63996d7aaa/corehq/apps/hqcase/case_helper.py#L192) in `case_helper`, but is currently inside a private function. You could move the dataclass out of the private method and then use it here.

Charl1996 replied:
Ah, thanks for pointing that out!

Charl1996 commented on corehq/apps/case_search/const.py:
```diff
@@ -59,15 +59,14 @@ def get_field_lambda(field_name):
                                          get_field_lambda('modified_on')),
 }
-SPECIAL_CASE_PROPERTIES = list(SPECIAL_CASE_PROPERTIES_MAP.keys())
+SPECIAL_CASE_PROPERTIES = list(SPECIAL_CASE_PROPERTIES_MAP.keys()) + ['commcare_case_copy']
```
I wanted to, but that results in a the following error:
> django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.

I know of a way to bypass this error, but it's not pretty and I'd much rather have the string literal than my bypass hack.

zandre-eng replied:
In that case, maybe just put a comment that this value should be the same as the constant? I'm just thinking about if we ever do change the string value and this line never gets updated.

Charl1996 committed:
Reuse UserDuck from case helper

Charl1996 committed:
Add comment

esoergel commented on corehq/apps/reports/filters/api.py:
```diff
@@ -197,3 +223,187 @@ def get(self, request, domain):
+
+class CaseCopier:
+    """A helper class for copying cases."""
+    COMMCARE_CASE_COPY_PROPERTY_NAME = "commcare_case_copy"
```
Nit: seeing this property without context I wasn't sure what it meant (it sounds like it might be a boolean).  Some other ideas after figuring it out:

* commcare_case_copy_of
* commcare_copy_of (we already know it's a case)
* commcare_original_case_id

esoergel commented on corehq/apps/reports/filters/api.py:
```diff
@@ -197,3 +223,187 @@ def get(self, request, domain):
+@require_permission(HqPermissions.edit_data)
+@requires_privilege(privileges.CASE_COPY)
+@location_safe
+def copy_cases(request, domain, *args, **kwargs):
```
nit: this file is `reports/filters/api.py`, which is used for providing lists of options to dropdown filters.  This functionality probably belongs elsewhere.

esoergel reviewed (commented):
I haven't read over the whole PR, just quickly skimmed it and added a couple comments, let me know if I'm off-base!

kaapstorm commented on corehq/apps/reports/standard/cases/filters.py:
```diff
@@ -149,6 +164,43 @@ def get_value(cls, request, domain):
+
+    def get_column_suggestions(self):
+        case_properties = get_flattened_case_properties(self.domain, include_parent_properties=False)
+        special_properties = [
+            {'name': prop, 'case_type': None, 'meta_type': 'info'}
+            for prop in SPECIAL_CASE_PROPERTIES if prop not in self.EXCLUDE_PROPERTIES
+        ]
```
It looks like `self.EXCLUDE_PROPERTIES` is almost all of `SPECIAL_CASE_PROPERTIES`. You could simplify this, and make what's happening here more obvious, if you just list the remaining included properties:
```suggestion
        special_properties = [
            {'name': prop, 'case_type': None, 'meta_type': 'info'}
            for prop in ('name', 'case_name', 'external_id')
        ]
```

kaapstorm commented on corehq/apps/reports/standard/cases/filters.py:
```diff
@@ -149,6 +164,43 @@ def get_value(cls, request, domain):
+    slug = "sensitive_properties"
+    label = gettext_lazy("De-identify options")
+    template = "reports/filters/sensitive_columns.html"
+    EXCLUDE_PROPERTIES = [
```
If this is only to be used in `get_column_suggestions()` you could move it into there.

kaapstorm commented on corehq/apps/reports/standard/cases/filters.py:
```diff
@@ -131,16 +139,23 @@ def get_column_suggestions(self):
         return case_properties + special_properties
 
+    @property
+    def exclude_special_case_properties(self):
```
Instead of copying code in `CaseListExplorerColumns` and `CaseSearchFilter`, you could drop this method from both classes (DRY), and replace the `special_properties` or `special_case_properties` variables with a function call to something like this:
```python
def get_special_case_properties(domain):
    from corehq.apps.reports.filters.api import CaseCopier

    props = SPECIAL_CASE_PROPERTIES + CASE_COMPUTED_METADATA
    if not toggles.COPY_CASES.enabled(domain):
        props -= [CaseCopier.COMMCARE_CASE_COPY_PROPERTY_NAME]
    return [
        {'name': prop, 'case_type': None, 'meta_type': 'info'}
        for prop in props
    ]
```

kaapstorm commented on corehq/apps/users/signals.py:
```diff
@@ -51,3 +57,4 @@ def connect_user_signals():
     couch_user_post_save.connect(sync_user_phone_numbers, dispatch_uid="sync_user_phone_numbers")
     commcare_user_post_save.connect(apply_correct_demo_mode,
                                     dispatch_uid='apply_correct_demo_mode')
+    commcare_user_post_save.connect(remove_test_cases, dispatch_uid='remove_test_cases')
```
So if the original mobile worker is deleted, then the copies of their cases are also deleted? Maybe add a comment here explaining that (and if I've misunderstood, please definitely add a comment).

Charl1996 replied:
@kaapstorm 
With "original mobile worker" I presume you mean "the owner of the original case"? A copied case is not deleted when the "original owner" is deleted, only when the owner of the copied case is deleted. The copied case is completely detached from the original case. 

This `remove_test_cases` only tags the copied cases belonging to this particular mobile worker as "deleted" when the user is deactivated. [This](https://github.com/dimagi/commcare-hq/blob/3e62cb7edab54575023b316b868e1ae5e8177d7e/corehq/apps/users/tasks.py#L398) is the celery task that executes when the mobile worker is deactivated.

kaapstorm replied:
When a mobile worker is deactivated, we already delete their cases, don't we? Are we doing something different here?

Charl1996 commented:
@esoergel 
> Does it belong in the SPECIAL_CASE_PROPERTIES list, then?

Honestly, I've been struggling with this question from the beginning, and I think you're right, it doesn't belong there. For some reason I thought this was also used to restrict users from setting case properties with these names, but turns out not.

Charl1996 commented on corehq/apps/data_interfaces/static/data_interfaces/js/case_management.js:
```diff
@@ -199,15 +273,25 @@ hqDefine("data_interfaces/js/case_management",[
             if ($select.length) {
                 $select.select2({
-                    placeholder: gettext("Search for users or groups"),
+                    placeholder: gettext(placeholderText),
```
> @Charl1996 can you confirm that this is what you intended?

Correct. (sorry, I missed Sravan's original question, so thanks for taking this up again, Norman.)

Charl1996 committed:
Remove gettext and use const from class

Charl1996 committed:
Remove const from SPECIAL_CASE_PROPERTIES

Charl1996 committed:
Remove class const and use reduced list

Charl1996 committed:
Add CaseCopier to case_helper.py

Charl1996 committed:
Move copy_cases endpoint from filters api to report views

Charl1996 committed:
Restrict only copy case prop in DD

Charl1996 committed:
Fix tests

Charl1996 committed:
Remove unused imports

Charl1996 committed:
Change success message

Charl1996 committed:
Support bulk case copy

Charl1996 committed:
Fix linting issue

Charl1996 committed:
Merge branch 'master' into cs/SC-2636-copy_cases_to_test_user

kaapstorm commented on corehq/apps/data_dictionary/models.py:
```diff
@@ -125,6 +124,10 @@ def get_or_create(cls, name, case_type, domain):
                 )
             except CaseProperty.DoesNotExist:
+                from corehq.apps.hqcase.case_helper import CaseCopier
+                if name in [CaseCopier.COMMCARE_CASE_COPY_PROPERTY_NAME]:
```
Nit
```suggestion
                if name == CaseCopier.COMMCARE_CASE_COPY_PROPERTY_NAME:
```

Charl1996 replied:
As I was writing this code I thought of doing the same thing, but changed my mind to "keep it generic" as in "just add to this list if you want some additional restricted property restricted". Looking at it now I realise I would probably look at this code in the future and think "why did this person do it this way?"...and maybe that's a good indication to do it differently.

kaapstorm commented on corehq/apps/data_dictionary/util.py:
```diff
@@ -207,9 +207,13 @@ def save_case_property(name, case_type, domain=None, data_type=None,
+            name=name, case_type=case_type, domain=domain
+        )
+    except ValueError as e:
+        return gettext(str(e))
```
Is the exception going to be translated?

Charl1996 replied:
After reading up a bit it seems like a bad idea to use `gettext` for use-cases like this. If translation is necessary, `ugettext_lazy` might be better, but the question is whether or not the exception _should_ be translated, in which case I think it might not be necessary (if fact, later in that same method we return an error without translating it, so maybe it's OK to not do that).

kaapstorm commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -609,37 +611,74 @@ def xform_management_job_poll(request, domain, download_id,
+        if self.is_copy_action:
+            action_text = "Copying"
+        else:
+            action_text = "Reassigning"
```
I think you need to translate these, because on line 644 you're using them as if they're already translated.

Charl1996 replied:
I'm always confused by string translations with variables. [This example](https://docs.djangoproject.com/en/4.2/topics/i18n/translation/#:~:text=This%20technique%20lets%20language%2Dspecific%20translations%20reorder%20the%20placeholder%20text.%20For%20example%2C%20an%20English%20translation%20may%20be%20%22Today%20is%20November%2026.%22%2C%20while%20a%20Spanish%20translation%20may%20be%20%22Hoy%20es%2026%20de%20noviembre.%22%20%E2%80%93%20with%20the%20month%20and%20the%20day%20placeholders%20swapped.) from the Django docs suggests that the string interpolation will be handled correctly without first translating the variable. I might be wrong?

kaapstorm replied:
Well spotted @millerdev !

https://github.com/dimagi/commcare-hq/pull/36633

kaapstorm reviewed (commented):
Just a few comments. Nothing major.

Charl1996 commented:
@kaapstorm 
> When a mobile worker is deactivated, we already delete their cases, don't we?

Not quite. Cases that belongs to a mobile user which is now deactivated will still be open (tested this on staging).

Charl1996 committed:
Update nit

Charl1996 committed:
Fix alert_user import
